Add new QNAP QSW uptime timestamp sensor

Noltari opened the pull request:
## Breaking change
QNAP QSW uptime seconds sensor has been replaced by an uptime timestamp sensor.

## Proposed change
Add new QNAP QSW uptime timestamp sensor.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Noltari committed:
qnap_qsw: sensor: add uptime timestamp

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

joostlek commented:
Why do we have 2 entities for this now? I prefer the date time one more since it doesn't write state every update

Noltari committed:
qnap_qsw: sensor: remove uptime_seconds

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

joostlek commented on homeassistant/components/qnap_qsw/sensor.py:
```diff
@@ -140,12 +143,12 @@ class QswSensorEntityDescription(SensorEntityDescription, QswEntityDescription):
         subkey=QSD_TX_SPEED,
     ),
     QswSensorEntityDescription(
-        translation_key="uptime",
```
Can we keep the old one around and deprecate that one properly? Reolink did a PR a few days back where it had to deprecate one, so maybe that's a good inspiration

Noltari committed:
Revert "qnap_qsw: sensor: remove uptime_seconds"

This reverts commit 1bdab4e3c5abf3ee19d0e67ab48ade082903b4cf.

Noltari committed:
qnap_qsw: sensor: deprecate uptime seconds entity

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
qnap_qsw: restore uptime_seconds to uptime

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

joostlek commented on homeassistant/components/qnap_qsw/sensor.py:
```diff
@@ -58,6 +63,17 @@ class QswSensorEntityDescription(SensorEntityDescription, QswEntityDescription):
     attributes: dict[str, list[str]] | None = None
     qsw_type: QswEntityType | None = None
     sep_key: str = "_"
+    value_fn: Callable[[str], datetime | float | int | str | None] = lambda value: value
```
```suggestion
    value_fn: Callable[[str], datetime | StateType] = lambda value: value
```

joostlek commented on homeassistant/components/qnap_qsw/strings.json:
```diff
@@ -51,8 +51,17 @@
         "name": "TX speed"
       },
       "uptime": {
-        "name": "Uptime"
+        "name": "Uptime seconds"
+      },
+      "uptime_timestamp": {
+        "name": "Uptime timestamp"
       }
```
I think uptime is still a valid and common name for the start up timestamp

Noltari committed:
qnap_qsw: add changes suggested by @joostlek

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

joostlek commented on homeassistant/components/qnap_qsw/strings.json:
```diff
@@ -52,7 +52,16 @@
       },
       "uptime": {
         "name": "Uptime"
+      },
+      "uptime_timestamp": {
+        "name": "Uptime timestamp"
       }
```
Hmm, something in me things uptime is viable for both, but I'm not sure what is best here

Noltari replied:
I'm happy with both, so I can change it to use uptime for both if you prefer that.

joostlek replied:
The only downside is that they'll both be called uptime so one will be appended `_2`

Noltari replied:
> The only downside is that they'll both be called uptime so one will be appended `_2`

Ah, you're right, and we won't be able to tell them appart for `uptime_seconds_deprecated`...

joostlek replied:
I've asked for a second opinion on this. 

But also keep in mind, people could've renamed both the name and entity id, so it could but even be recognizable

Noltari committed:
qnap_qsw: only warn if deprecated sensor used in automations/scripts

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into qnap-qsw-uptime-timestamp

Noltari committed:
qnap_qsw: fix unique_id deprecation comparisson

The Unique ID was accidentally changed on 3caffa4d:
- Before 3caffa4d: the unique ID ends with "_uptime".
- After 3caffa4d: the unique ID ends with "_uptime_seconds".

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into qnap-qsw-uptime-timestamp

Noltari committed:
Merge branch 'dev' into qnap-qsw-uptime-timestamp

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into qnap-qsw-uptime-timestamp

Noltari commented:
Failure in codecov not related to changes:
https://github.com/home-assistant/core/pull/122589/checks?check_run_id=28777461908

Noltari committed:
Merge branch 'dev' into qnap-qsw-uptime-timestamp

joostlek commented on homeassistant/components/qnap_qsw/sensor.py:
```diff
@@ -337,6 +355,46 @@ async def async_setup_entry(
+                        hass,
+                        DOMAIN,
+                        f"uptime_seconds_deprecated_{entity_id}_{item}",
+                        breaks_in_ha_version="2025.2.0",
```
```suggestion
                        breaks_in_ha_version="2025.5.0",
```

joostlek committed:
Update homeassistant/components/qnap_qsw/sensor.py

joostlek commented on homeassistant/components/qnap_qsw/sensor.py:
```diff
@@ -337,6 +355,46 @@ async def async_setup_entry(
             )
             entities.append(QswSensor(coordinator, _desc, entry, port_id))
 
+    # Can be removed in HA 2025.2.0
```
```suggestion
    # Can be removed in HA 2025.5.0
```

joostlek committed:
Update homeassistant/components/qnap_qsw/sensor.py

joostlek committed:
Update homeassistant/components/qnap_qsw/sensor.py

Noltari commented:
Thanks @joostlek!
